### PR TITLE
Improve switch statement complexity measure

### PR DESCRIPTION
**Pull request by garnett**
Fixes issue described in https://github.com/realm/SwiftLint/issues/461

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift [hunk 1]
```diff
@@ -36,8 +40,7 @@ public struct CyclomaticComplexityRule: ASTRule, ConfigProviderRule {
             return []
         }
 
-        let substructure = dictionary["key.substructure"] as? [SourceKitRepresentable] ?? []
-        let complexity = measureComplexity(substructure) + 1
+        let complexity = measureComplexity(file, dictionary: dictionary) + 1
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift]: I think this `+ 1` is wrong. See my comment here: https://github.com/realm/SwiftLint/issues/461#issuecomment-178115875

File: Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift [hunk 2]
```diff
@@ -51,25 +54,50 @@ public struct CyclomaticComplexityRule: ASTRule, ConfigProviderRule {
...
+
+        let c = (file.contents as NSString)
+            .substringWithByteRange(start: bodyOffset, length: bodyLength) ?? ""
+
+        let fallthroughCount = c.componentsSeparatedByString("fallthrough").count - 1
+        return complexity - fallthroughCount
     }
 
     private let complexityStatements = [
         "source.lang.swift.stmt.foreach",
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift]: it'd be nice to get these into SourceKitten much like `SyntaxKind` and `SwiftDeclarationKind`.

File: Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift [hunk 3]
```diff
@@ -20,13 +21,16 @@ public struct CyclomaticComplexityRule: ASTRule, ConfigProviderRule {
...
+            "switch code {\n case 0: fallthrough\ncase 0: return 1\ncase 0: return 1\n" +
+            "case 0: return 1\ncase 0: return 1\ncase 0: return 1\ncase 0: return 1\n" +
+            "case 0: return 1\ncase 0: return 1\ndefault: return 1}}"
         ],
         triggeringExamples: [
             "func f1() {\n  if true {\n    if true {\n      if false {}\n    }\n" +
-                "  }\n  if false {}\n  let i = 0\n\n  switch i {\n  case 1: break\n" +
-                "  case 2: break\n  case 3: break\n  default: break\n  }\n\n" +
-                "  for _ in 1...5 {\n    guard true else {\n      return\n    }\n  }\n}\n"
+            "  }\n  if false {}\n  let i = 0\n\n  switch i {\n  case 1: break\n" +
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift]: we lost the indentation here

File: Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift [hunk 4]
```diff
@@ -20,13 +21,16 @@ public struct CyclomaticComplexityRule: ASTRule, ConfigProviderRule {
         description: "Complexity of function bodies should be limited.",
         nonTriggeringExamples: [
             "func f1() {\nif true {\nfor _ in 1..5 { } }\nif false { }\n}",
-            "func f3() {while true {}}",
+            "func f(code: Int) -> Int {" +
+            "switch code {\n case 0: fallthrough\ncase 0: return 1\ncase 0: return 1\n" +
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift]: these should be indented to indicate that they're additions to the string above.

**Comment by jpsim** [comment]
This could use a changelog entry, otherwise I think it's on the right track.

File: Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift (hunk 2 above)
Comment (garnett) [inline Source/SwiftLintFramework/Rules/CyclomaticComplexityRule.swift]: Looks like good idea to me :+1: maybe I will do that too!

**Comment by garnett** [comment]
@jpsim non related to this particular PR:

Do you think it is worth extracting things like `dictionary["key.substructure"] ` into dictionary extension to provide syntax like `dictionary.substructure`? Maybe this is even more related to SourceKitten.

**Comment by jpsim** [comment]
> Do you think it is worth extracting things like dictionary["key.substructure"] into dictionary extension to provide syntax dictionary.substructure? Maybe this is even more related to SourceKitten.

Yeah, there's a WIP PR on SourceKitten to use proper types rather than dictionaries, which would help greatly with that.

**Comment by jpsim** [comment]
This looks good to me, thanks @garnett!

**Comment by jpsim** [comment]
Merging in #474.

**Pull request closed without merging**